src/Coordination/KeeperStorage.cpp

```diff
@@ -166,54 +166,132 @@ KeeperStorage::ResponsesForSessions processWatchesImpl(
 }
 
 // When this function is updated, update CURRENT_DIGEST_VERSION!!
-uint64_t calculateDigest(std::string_view path, std::string_view data, const KeeperStorage::Node::Stat & stat)
+uint64_t calculateDigest(std::string_view path, const KeeperStorage::Node & node)
 {
     SipHash hash;
 
     hash.update(path);
 
-    hash.update(data);
+    auto data = node.getData();
+    if (!data.empty())
+    {
+        chassert(data.data() != nullptr);
+        hash.update(data);
+    }
+
+    hash.update(node.czxid);
+    hash.update(node.mzxid);
+    hash.update(node.ctime());
+    hash.update(node.mtime);
+    hash.update(node.version);
+    hash.update(node.cversion);
+    hash.update(node.aversion);
+    hash.update(node.ephemeralOwner());
+    hash.update(node.numChildren());
+    hash.update(node.pzxid);
+
+    auto digest = hash.get64();
+
+    /// 0 means no cached digest
+    if (digest == 0)
+        return 1;
```

Comment:
I guess this is to return any value that's different from 0, right?

Reply:
yes
I had to do this in the end because I couldn't use `has_cached_digest` bool with, for example, `mtime`.
Caching digest can be modified for a node that is read by snapshot so we can have a race in that case.